### PR TITLE
Add DimensionDictionary, which handles identification of dimensions into categories

### PR DESCRIPTION
**Pull request by Ivorforce**
This works similar to BiomeDictionary, just for dimensions.

Currently it's near impossible to identify dimensions for any practical purpose. 
The addition of this dictionary answers questions such as:

* Should I spawn my structures/entities/ores in dimension X?
* Does dimension X belong to mod X? (Assuming the mods adds its own category)
* Is dimension X the well-known dimension by the name of 'Y'? (semi-ID)
* Is the player currently on a different planet? 
* Does dimension X allow magic/enchantments/mechanic y

### CODE REVIEW

**Comment by heldplayer** [comment]
Looks good, would be a nice addition to Forge :+1:

**Comment by GotoLink** [comment]
In its current state, your suggestion would not help answer any of those questions.
I suggest you enforce "type" exclusions, at the very least.
For example excluding "REAL" subtypes from the "UNREAL" subtypes...or actually registering some subtypes to "REAL".
Also, why is "HELL" a subtype of "UNDERGROUND" ? (outside of religious interpretation, the "HELL" doc wouldn't imply "UNDERGROUND"). Same goes for "ENDER" and "SKY".

**Comment by Abastro** [comment]
I suggest adding some hook to the DimensionManager.

**Comment by McJty** [comment]
I would love to see this in Forge as well. Will be very handy for RFTools dimensions in combination with Ivorius structure building mod

**Comment by ThePsionic** [comment]
@GotoLink I can get into the "ENDER" and "SKY" connection since the End is technically a Sky biome. I do agree with you that "HELL" and "UNDERGROUND" do not have any connection; in that case I would have named the latter "NETHER" or something.

**Comment by Ivorforce** [comment]
ThePsionic, I just made some substantial changes, which addresses this issue. Reload the page :)

**Comment by ThePsionic** [comment]
@Ivorforce looks good now, no opposition from me anymore :P

**Comment by Ivorforce** [comment]
@GotoLink Your last concern has been addressed in the same commit as Psionic's.
I don't know what subtypes you'd expect for 'REAL', to be honest. If you have suggestions, go ahead and say it :P
Lastly, the exclusions aren't a thing to keep it more simple. Technically you'd want a nullable restricted number restricting some of the types that are mutually exclusive.
That would overcomplicate the process of using the dictionary unnecessarily though, especially since you can't always know what's completely logically exclusive (nether <-> end <-> earth? Who knows...). I'd rather not get into the whole thing and just work with simple binary string mapping.

**Comment by GotoLink** [comment]
Well "EARTH" feels like it should have physics in it, to me...thus subtype of "REAL".

Though my main issue is with the general design of it all.
Mixing physics, space constraints, and design purpose in single names, that just feels unpractical and inefficient (i mean, 'dimensionMatchesType' look-up time is not going to be any good in this design). Not only the lexical issues i have already pointed at.

**Comment by heldplayer** [comment]
An alternative could be to actually attach the data to the provider, which would make it easier for people that add dimensions to add the tags. We wouldn't have a DimensionDictionary but instead the properties are put on the provider itself and it can be asked if it has a certain tag.

**Comment by XCompWiz** [comment]
Long comment is long. Apologies.  I just have a good deal to say apparently.

I agree with held; add this to the world provider, not as a global registration system.  This permits better flexibility, particularly with how dimensions are stored and generated.
The drawback is you need an instance of the world to be able to ask, but the target use case is generation at which point we'd have such an object.
Mystcraft could not use this system as it stands well, as it doesn't know what a world is like until it's started adding things.  It will already have trouble determining such categories with the data for the world fully loaded.

Strictly speaking, most types should not have subtypes.
For example, just because all existing examples of a boss fight dimension are finite, this does not mean all instances of a boss fight dimension need to be (perhaps a boss fight dimension set in an infinite maze).
Since multiple categories are possible, it simplifies matters to allow for combinations which are outside of any assumed hierarchy.
Unless the definition of something implies a category (imaginary is inherently unreal), then treating it as a subcategory of something else might break it's potential uses.  Having such high level categories is somewhat useful, though, so there I'd accept them.

Secondly, I'd argue there is a better way to categorize worlds as having or lacking a top and bottom layer.  You have 4 categories covering all permutations of two state variables.  Better to break it down into those states.  If you want explicit On/Off for those states you can have them and still have four tags.  This makes a difference if we later add a third On/Off state, which would cause 8 categories in your current arrangement, only 6 with explicit off states, and 3 with only on states.

Taking that one step further, you can really convert this into a "stateful" system (is/has), allowing a query system for dimensions.  Ask the world provider for the dim if the dimension "has roof" or "is empty (void)" as a string constant.  This even allows inventing new strings which can be later added to Forge or not depending on uses (a mod can use for it's own dimensions "isUserConstructed" or so).

The Earth, Hell, and Ender categories are oddly both too specific and uninformative. What makes a dimension like the Ender?  Endermen? An Ender Dragon? End Stone?
The first you'd be better off querying spawn options. The second... "Boss" might be useful; spawning might work too.  The last you can just check the existing terrain generated.
Break those up into actual categorizations, don't name a category off of existing dimension types.

tl;dr
* Make a part of the world provider, not a global registration system.
* Limit the use of types as subtypes to inherent by dictionary definition.
* Earth, Hell, and Ender are bad categories.
* Make this a query system for state ("Roof", "VoidCover", "Fauna", "Imaginary") or at least change the categories to be purely single states ("Surface" is TWO states).

**Comment by Ivorforce** [comment]
@GotoLink The 'earth' category is meant as a theming category. Simple example: If the dimension embodies a *dream* of the *earth*, the categories are *DREAM* and *EARTH*. Subtyping this doesn't really make sense here.
I hope the inefficiency complaint is a joke. This runs over and queries 1, 2 or if somebody goes absolutely ballistic maybe even 100 strings, but it will never be slow. You don't use this 300 times a tick.
Impractical, maybe, but it's hard to really define what dimensions are. A tag system is generally applicable, versatile, easy to use and modular. At the very least it's better than what we have now, which is nothing.

@XCompWiz Hot damn, welp, let me break this down myself. x)

1) **Make a part of the world provider, not a global registration system**
-- First off - yes, this makes sense for dynamic dimensions. My thoughts when making this system currently were that dynamic dimensions can register and unregister dynamically, on construction and, likewise, destruction. I guess as a part of Forge, rather than my personal mod though, an interface for the world provider is also viable.
This will, of course, not work for vanilla dimensions, so the wrapper methods will stay as-is, and I think I'll also keep the static assignment for simple registration of the default static dimensions. Downside is that the dynamic dimensions won't be modular, but then again, that might not be that important.

2) **Limit the use of types as subtypes to inherent by dictionary definition.**
Well, I guess this is really just a discussion that breaks down to every individual case. I guess the 'boss arena' thing doesn't have to be finite, so I'll change that too.

3) **Make this a query system for state ("Roof", "VoidCover", "Fauna", "Imaginary") or at least change the categories to be purely single states ("Surface" is TWO states). **
We actually went over this in the channel. My argument for the current approach was that each of these 4 permutations is rather special in definition, and can even be named, since the whole nature of this question is interestingly specific. 'Is there, theoretically, terrain above/below the dimension?' defines 4 very distinct states, as named in the PR. I don't think a 3rd category will really be added, as I can't think of any, but who knows I guess. It's not that important to me; I can see the argument for individual booleans and if you want me to I'll just change that around again.

I'm not sure where you're going with states / queries though. The dictionary is just about representing dimension states, and if you just pass in string constants and return booleans, you might as well just stay with a dictionary / Set<String> approach. Perhaps you can enlighten me in that regard though?

The problem with assuming each query yields a boolean is that not every type is boolean. Sure, you can ask if a type is 'not the earth', but 'not the earth' as an individual type doesn't really make sense - which, looking at 'has roof' vs 'has no roof' would have to be the case there (two individual types).

** Earth, Hell, and Ender are bad categories.**
Imho, checking every single value for likeliness of being some type I'm making up in my head is overly complicated and unusable. Not only would it be slow, it would give me only some approximate values where the intention of the author should be quite clear. The nether is clearly 'hell' themed, and the end 'end' themed - hell, its name is 'end', it has 'endermen', 'endstone' and an 'enderdragon'. I can think of plenty of mods, structures, or entities that can make use of this by wanting to expand only these specific types of dimensions, or add their own variations. 
Likewise, mods will also want to add their own specific brands to dimensions, and other modders will want to use these to specifically target these types of dimensions. Personally, I can make use of all these kinds of tags, really. :P

You can argue that those are very specific - which they are; pretty much by definition - but we can always break this down to more supertypes if you can think of any fitting ones.

**Comment by XCompWiz** [comment]
There's very little value to a specific "Is the Overworld" category; and that's all Earth would ever be useful for; since we can check for that anyway.  I'd want to know if "has plants" if I'm a flora and fauna mod, for example.
Granularity to that information is important.  Unless I'm creating an Overworld clone, the Overworld category isn't correct form my dimension.  Using the category as anything less strict will produce funny behaviour.

"I don't think a 3rd category will really be added" - Always assume that the needs will exceed the initial design, and leave the design open for more things later.
Alternatively: "Never code yourself into a corner."
Example 3rd category:  Describe the middle section between the roof and the void cover. ex:  Solid middle, dense middle, sparse middle, empty middle.

If you have both ways (query by world provider AND query by global system) you are begging for bugs.
If you go with the global registration system then Mystcraft/dynamic dimension mods probably won't be able to use this well.  At least for Mystcraft the dimension information isn't built until right before generation occurs.

How is Planet_Surface different from Surface?  Would a Planet category not convey the same meaning with more generality to it? Is Planet Cave World not a thing?

Non-binary state is a very good point, and at that level this approach fares better. :)

**Comment by Ivorforce** [comment]
I'm assuming by 'overworld category' you mean the 'earth' type?
Well, you can always hardcode yourself to dimension==0, but that would just be as hardcode-y in my opinion, and wouldn't work for mods. Even if you think anything less than handling it completely strictly is acceptable for those, they'll still be useful, from a query standpoint - if mod X adds dimension Y, it can simply add its own tag for this dimension and it's instantly clear it's what I'm looking for, rather than having to do some weird class analysis.
Anyway, if you want a specific set of types, I propose you write them down for discussion - I'm by no means omniscient in the matter of how dimensions may want to represent themselves, and I'm sure you have a good amount of insight in that manner :P

3rd category - it's a good idea, but nonetheless you wouldn't have to combine it with the first two to create more permutations. Whether or not you use the current system as yours is, as they represent exactly the same states, arbitrary, so from a definition perspective it doesn't matter which we choose. I'll just change it to your suggestion and then we're good on that front - as mentioned, this isn't all that important to me :P

Obviously, I won't go solely with the static registration system at this point, but - if we keep it nonetheless, we can avoid sticking an arbitrary interface to vanilla classes. I'm sure Lex will be happy if he can keep the patches to a minimum. I'm not exactly sure where bugs would come in, except potential values getting lost if somebody tries to register for a dimension that implements the interface.

A dimension can clearly have a surface without logically being a planet. Planet Cave World is an argument though, so I'm not sure x)

File: src/main/java/net/minecraftforge/common/DimensionDictionary.java [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+     * @param provider The dimension's provider.
+     * @return The dimension's types.
+     */
+    public static Set<String> getDimensionTypes(WorldProvider provider)
+    {
+        if (provider instanceof Handler)
+            return ((Handler) provider).getDimensionTypes();
+            
+        Set<String> types = dimensionTypes.get(provider.dimensionId);
+        return types != null ? SET_UNCATEGORIZED : Collections.<String>emptySet();
```
Comment (XCompWiz) [inline src/main/java/net/minecraftforge/common/DimensionDictionary.java]: If types != null returns SET_UNCATEGORIZED
if types == null, returns Collections.<String>emptySet();

Neither case actually returns types. :P

File: src/main/java/net/minecraftforge/common/DimensionDictionary.java [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+    public static boolean dimensionMatchesType(WorldProvider provider, String type)
+    {
+        Set<String> dimTypes = getDimensionTypes(provider);
+
+        Queue<String> curTypes = new ArrayDeque<String>();
+        curTypes.add(type);
+
+        String curType;
+
+        while ((curType = curTypes.poll()) != null)
```
Comment (XCompWiz) [inline src/main/java/net/minecraftforge/common/DimensionDictionary.java]: Why is the type string put into a collection only to pull it out of the collection?

File: src/main/java/net/minecraftforge/common/DimensionDictionary.java (hunk 1 above)
Comment (Ivorforce) [inline src/main/java/net/minecraftforge/common/DimensionDictionary.java]: Did I actually replace the wrong value? lol

File: src/main/java/net/minecraftforge/common/DimensionDictionary.java (hunk 2 above)
Comment (Ivorforce) [inline src/main/java/net/minecraftforge/common/DimensionDictionary.java]: Because it was faster to code. You happy now? :P

**Comment by XCompWiz** [comment]
Nice edits. :)
We need more input from people who would use this system. What do people going generation in dimensions want to know about that dim?

**Comment by GotoLink** [comment]
I would really like some flags such as
'isGroundPathable', 'isAirFlyable'
for mob spawning;
'hasVillages', 'hasPlants', 'hasOres'
for generation/mob behavior;
'hasLiquids', 'hasGravity', 'hasPowerFlow'
for technical blocks/entity behavior.
Patching WorldProvider would surely help achieve that.

**Comment by Ivorforce** [comment]
@XCompWiz Well, personally I have a structure mod that expands upon the world generation. Every structure itself determines where it wants to spawn, so theoretically any type of tag can be useful.

A lot of the selection can be handled by the biome dictionary, but for advanced selection tags such as this can be useful:
'Does the dimension really want to accept structures at all?' -> Hence ABSTRACT + FINITE and similar tags.
'What is the general shape of the terrain?' -> Hence the bottom/top limit tags
'What is the existential context of the dimension?' -> Hence REAL/UNREAL tags, and anything mods may extend upon. This includes, for example, structures that want to spawn solely and specifically in, say, Twilight Forest dimensions.

Anything else dimensions can find to split themselves into logical groups would be useful, too. That's why I added the 'PLANET_SURFACE' tag, for example. The current list of tags is enough for me to make meaningful distinctions already, though, granted mods add their own tags to at least be identifiable by a semi-ID.

I don't... entirely agree with @GotoLink. Most of that stuff is, in my opinion, biome specific (except perhaps hasGravity) rather than necessarily spanning the whole dimension.

**Comment by Ivorforce** [comment]
@XCompWiz Since it has recently come up for me again, any more thoughts on this?

Default tags can always be changed, or simply ignored in favour of more appropriate tags. What I'd really like personally though is for the system itself to be approved and pulled, no matter of the specific content. As mentioned, I merely added those I'd personally be interested in for the moment.

Lex pointed to you to make the decision, so I'd love to hear your current stance.

**Comment by XCompWiz** [comment]
As a note, once added, tags cannot be changed. Once in Forge, things remain static and must be supported at least until the next major version of Minecraft.
Ignoring tags as we add better ones undermines the system, but is somewhat of an option.

My stance is still no to this implementation. I don't think a static tagging system is flexible enough, and building it so externalized from the dimension instances inherently lends itself to misuse.  The current design doesn't support dynamic dimensions almost at all, nor does it allow for properties to change over the lifetime of a dimension.

I've recently proposed (in response to some feedback and discussion with others who would use the information) a property-based system where you could request properties about a dimension by string identifiers.  Some default ones could be defined (so we can always have some baseline to work with)  and custom ones could be negotiated between mods.  We can add more over time, but want to be really certain of the ones we define in Forge.  Allowing arbitrary strings permits mods to have their own without Forge having to explicitly support the property.

I actually think that biome and world properties should all be filtered by the world provider, so that it can enforce a consistent theme and properly control the dimension.  Unfortunately, there are lots of places where information is taken directly from the biome at a location, completely eliminating opportunities for more elaborate behavior.

Anyway, I should make a more formal proposal for this system (and probably wright the code for it) as a PR.  I say no to this one. Sorry, and sorry for leaving this hanging for so long.

**Comment by Ivorforce** [comment]
Right. I'm officially closing this one then, in hopes of your proposed system getting made rather soon.

As a clarification - The static tag system was thought up with the static dimensions in mind - namely the 3 vanilla ones we have. However, keeping modding support and future changes open, I agree it might be best to transition those to an instance based system themselves - preferably with a hookable event.

I'm still not sure what would differentiate your system from the here supposed handlers [apart from being a set rather than returning one], but hopefully that will change when you propose your version.

**Comment by Ivorforce** [comment]
@XCompWiz Any ETA on this? It's been like a year and a half...

**Comment by XCompWiz** [comment]
It's only been a year as of right now. :P

But your points stands, it has been a while.  The concept I sketched above is still what I'd support and am willing to make as PR to Forge, but I have a lot of things to catch up on first.  If anyone is interested in working with me on making this PR so we can get it done sooner let me know.

**Comment by srs-bsns** [comment]
I don't think I am able to help you craft this feature, but I fully support it.

**Comment by cpw** [comment]
Before you start xcomp, there's been some changes in dimension handling in 1.9+ that might affect your view if this.

**Comment by XCompWiz** [comment]
Of course. :D I'll have to return to it after updating, I expect.

**Comment by Ivorforce** [comment]
@XCompWiz Any news on this? Otherwise I'd like to reopen this soon. I am still very interested in the dictionary...
